File: Read/Discrete_Mathematics_Using_a_Computer/C1.Intro/C1.ml
open Big_int

(* helpers *)

let rec take n l =
    if (l == []) then [] else
    match n with
    | 0 -> []
    | _ -> List.hd l :: take (n-1) (List.tl l)

let rec drop n l =
    if (l == []) then [] else
    match n with
    | 0 -> l
    | _ -> drop (n-1) (List.tl l)

let rec zipWith f l1 l2 =
    if ([] == l1 || [] == l2)
        then []
        else (f (List.hd l1) (List.hd l2)) :: zipWith f (List.tl l1) (List.tl l2)
    
let zip l1 l2 = zipWith (fun a b -> (a, b)) l1 l2

let list_of_string s =
    let rec tolist i acc =
        match i with
        | 0 -> s.[i] :: acc
        | _ -> tolist (i-1) (s.[i] :: acc)
    in
        tolist (String.length s - 1) []

let rec range i j =
    match (i > j) with
    | true -> []
    | false -> i :: range (i + 1) j

let compose f g x = f (g x)

let contains c s = String.contains s c

(* chapter *)

let int_types = (0, Big_int.zero_big_int)

let bigint = Big_int.zero_big_int

let int_exp = 9*9
let integer_exp = string_of_big_int (power_big_int_positive_big_int (big_int_of_int 9) (big_int_of_int 100))

let int_div = 10 / 2
let float_div = 10.0 /. 2.0

let float_math = 10.0 +. 2.1 /. 5.6 -. 1.0 *. 2.0 ** 5.0

let bools = not (true && false || true)
let comparison = (1 < 2 && 2 < 3) && (3 >= 2 && 2 >= 1) != false

let c = 'A'
let _C = Char.uppercase c

let s = "abcdef"
let _S = String.uppercase s
let _S' = String.concat "" (List.map (fun c -> String.make 1 (Char.uppercase c)) (list_of_string s))
let _S'' = String.map (fun c -> Char.uppercase c) s

let strings = String.concat "" ["abc";"def";"xyz"]

let tup = (1, "string")
let fstTup = fst tup
let sndTup = snd tup

let num_list = [1,2,3,4,5]
let num_lists = [num_list, num_list, num_list]
let num_lists' = List.map (fun _ -> num_list) [1;2;3]

let cons = 1 :: 2 :: 3 :: 4 :: 5 :: []
let twolists = cons @ cons

let double x = x * x
let double' = function x -> x * x
let double'' = fun x -> x * x

let type_signature (x : int) = x

exception Wrecked

let pattern_matching x =
    match x with
    | 0 -> false
    | 1 -> true
    | _ -> raise Wrecked

let pattern_matching' = function
      0 -> false
    | 1 -> true
    | _ -> raise Wrecked

let match_tuple (x,y) = x == y

let twice f a = f (f a)
let twice_example = twice (fun a -> a + a) 5

let quadratic a b c =
    let d = sqrt (b**2.0 -. 4.0*.a*.c) in
    let x1 = ((b*.(-1.0)) +. d) /. (2.0*.a) in
    let x2 = ((b*.(-1.0)) -. d) /. (2.0*.a) in
    (x1,x2)

let list_index_example = List.nth [1;2;3;4;5] 2
let take_example =
    let l = [1;2;3] in
    (take 2 l, take 0 l, take 4 l)

let drop_example =
    let l = [1;2;3] in
    (drop 2 l, drop 0 l, drop 4 l)

let map_example = String.map (fun c -> Char.uppercase c) "the cat and dog"

let zip_example = zip [1;2;3] ['a';'b';'c']
let zipWith_example = zipWith (fun a b -> a + b) [1;2;3] [1;2;3]

let foldl_example = List.fold_left (fun acc x -> x + acc) 0 (range 1 10)
let foldr_example = List.fold_right (fun acc x -> x + acc) (range 1 10) 0

let compose_example = compose Char.uppercase Char.lowercase 'c'

(*let compose_list = (compose (::) (fun x -> Char.uppercase x) 'a') "bc"*)

type color = Red | Yellow | Blue

type booL = False | True

type animal = Cat of string | Dog of string | Rat

type ('a, 'b) animal' = Cat' of 'a | Dog' of 'b | Rat'

type 'a maybe = Nothing |Just of 'a

exception MaybeException

let fromJust = function
      Just a -> a
    | _ -> raise MaybeException

let maybePlus a b =
    if (a == Nothing)
        then b
    else if (b == Nothing)
        then a
    else
        Just ((fromJust a) + (fromJust b))
    

(* exercises *)

let ex1 =
    zip
    [
        [true && false];
        [true || false];
        [not false];
        [3 <= 5 && 5 <= 10]
    ]
    (range 1 10)

exception WhoCares

let ex2 () = raise WhoCares

let ex3 c =
    match c with
    | 'a' -> true
    | _ -> false

let ex4 s =
    match s with
    | "hello" -> true
    | _ -> false

let ex5 s =
    let len = String.length s in
    if (len > 0)
        then
            if (s.[0] == ' ')
                then String.sub s 1 (len-1)
                else s
        else
            s

exception ToBoolException
       
let ex6 =
    let toBool = function
          0 -> false
        | 1 -> true
        | _ -> raise ToBoolException
    in
    List.map toBool

let ex6_example = ex6 [0;1;0;1;0;1;1;1]

let ex7 = contains '0'
let ex7_example = ex7 "Aleph"
let ex7_example' = ex7 "Aleph0"

let ex9_addjust = zipWith maybePlus
let ex9_addjust_example = ex9_addjust [Just 2; Nothing; Just 3] [Nothing; Nothing; Just 5]

type ex10 = Metal1 | Metal2 | Metal3 | Metal4 | Metal5 | Metal6

type 'a ex11 = Coin of 'a

type stuff = Bool of bool | Char of char | Int of int 
type ('a, 'b, 'c, 'd) tup4 = Tup4 of 'a * 'b * 'c * 'd

let ex14_quadratic a b c =
    let discrim = b**2.0 -. 4.0*.a*.c in
    let quad op = (op (b*.(-1.0)) (sqrt discrim))/.(2.0*.a) in
    if (discrim < 0.0)
        then Nothing
        else Just (quad (-.), quad (+.))

let bitOr b1 b2 =
    if (b1 == 0 && b2 == 0)
        then 0
        else 1

let bitAnd b1 b2 =
    if (b1 == 1 && b2 == 1)
        then 1
        else 0

let bitwiseAnd = zipWith bitAnd
let bitwiseOr = zipWith bitOr

let ex20 = List.map (fun a -> fromJust a) (List.filter (fun a -> a != Nothing) [Just 3;Nothing;Just 4])

File: Read/DiscreteMathematicsUsingAComputer/C1.Intro/C1.hs
import Data.List
import Data.Ratio
import Data.Char

int_bounds :: (Int, Int)
int_bounds = (minBound :: Int, maxBound :: Int)

int_types :: (Int, Integer)
int_types = (1 :: Int, 1 :: Integer)

int_exp = 9^9
integer_exp = 9**100

int_div = 10 `div` 2
double_div = 10 / 2

float_math = 10.0 + 2.1 / 5.6 - 1.0 * 2.0 ** 5.0

ratio = 2/3 :: Ratio Integer
denomRatio = denominator ratio
numRatio = numerator ratio

bools = not (True && False || True)
comparison = ((1 < 2 && 2 < 3) && (3 >= 2 && 2 >= 1)) /= False

c = 'A'
_C = toUpper c

s = "abcdef"
_S = map toUpper s

strings = "abc" ++ "def" ++ "xyz"

tup = (1, "string")
fstTup = fst tup
sndTup = snd tup

num_list = [1,2,3,4,5]
num_lists = replicate 3 num_list

alphabet = ['a'..'z']
alphabet' = map toUpper alphabet
base10 = [0..10]

cons = 1 : 2 : 3 : 4 : 5 : []
twolists = cons ++ cons

list_comprehension = [ x * x | x <- [1..10] ]
list_comprehensions = [ (x, y) | x <- list_comprehension, y <- list_comprehension ]

double x = x * x
lambda_double = \x -> x * x

type_signature :: Int -> Int
type_signature int = int

pattern_matching 0 = False
pattern_matching 1 = True

pattern_matching' x =
    case x of
        0 -> False
        1 -> True

pattern_matching'' x
    | x == 0 = False
    | x == 1 = True

match_tuple (x,y) = x == y

twice :: (a->a) -> a -> a
twice f a = f (f a)

quadratic a b c =
    let d = sqrt (b^2 - 4*a*c)
        x1 = (-b + d) / (2*a)
        x2 = (-b - d) / (2*a)
    in (x1,x2)

list_index_example = [1..5] !! 2

take_example =
    let l = [1,2,3] in
    (take 2 l, take 0 l, take 4 l)

drop_example =
    let l = [1,2,3] in
    (drop 2 l, drop 0 l, drop 4 l)

map_example = map toUpper "the cat and dog"

zip_example = zip [1,2,3] "abc"
zipWith_example = zipWith (+) [1,2,3] [1,2,3]

foldl_example = foldl (+) 0 [1..10]
foldr_example = foldr (+) 0 [1..10]

compose_example = (toUpper . toLower) 'c'
compose_list = ((:) . toUpper) 'a' "bc"

data Colour = Red | Yellow | Blue

data Animal = Cat String | Dog String | Rat

data Animal' a b = Cat' a | Dog' b | Rat'

-- exercises

ex1 = zip [True && False, True || False, not False, 3 <= 5 && 5 <= 10, 3<= 20 && 20 <= 10, False == True, 1 == 1, 1 /= 2, 1 /= 1] [1..]

ex2 =
    [
        [null [x | x <- [1,2,3], False]],
        [not (x && y) | x <- [False, True], y <- [False, True]],
        [x || y | x <- [False, True], y <- [False, True], x /= y]
    ]

ex3 c
    | c == 'a' = True
    | otherwise = False 

ex4 "hello" = True
ex4 _ = False

ex5 s@(sh:st)
    | sh == ' ' = st
    | otherwise = s

ex6 = map toBool
    where
        toBool n
            | n == 0 = False
            | n == 1 = True
            | otherwise = error "toBool: must be 0 or 1"

ex6_example = ex6 [0,1,0,1,0,1,1,0]

ex7 = any (=='0')
ex7_example = ex7 "Aleph"
ex7_example' = ex7 "Aleph0"

data List a = Cons a | Nil deriving (Show)

data Foldr op base list = Foldr op base list deriving (Show)

data Op a b = Max a b deriving (Show)

--foldr_unfold = Foldr max 0
foldr_unfold = Foldr Max 0 (Max 1 (Max 5 (Max 3 0)))

instance (Num a) => Num (Maybe a) where
    (+) Nothing b = b
    (+) a Nothing = a
    (+) (Just a) (Just b) = Just (a + b)

ex9_addJust = zipWith (+)

ex9_addJust_example = ex9_addJust [Just 2, Nothing, Just 3] [Nothing, Nothing, Just 5]

data Ex10 = Metal1 | Metal2 | Metal3 | Metal4 | Metal5 | Metal6 deriving (Eq)

data Ex11 a = Coin a

data Stuff = BOOL Bool | CHAR Char | INT Int

type Tup4 a b c d = (a,b,c,d)

ex14_quadratic a b c =
    case (discrim < 0) of
        True -> Just (quad (-), quad (+))
        False -> Nothing
    where
        discrim = b^2 - 4*a*c
        quad op =(-b `op` sqrt discrim)/(2*a)

bitOr :: Int -> Int -> Int
bitOr 0 0 = 0
bitOr _ _ = 1

bitAnd :: Int -> Int -> Int
bitAnd 1 1 = 1
bitAnd _ _ = 0

bitwiseAnd = zipWith bitAnd
bitwiseOr = zipWith bitOr

ex16 = bitwiseAnd

ex20 = map (\(Just a) -> a) $ filter (/= Nothing) [Just 3, Nothing, Just 4]

ex21 n = filter (>n)

ex22 ints int = map fst $ filter (\(f,s)->s==int) $ zip [1..] ints

-- tired going to pass out
